spec/sailfish_iptables_dbus_interface_description.xml

```diff
@@ -0,0 +1,464 @@
+"<!DOCTYPE node PUBLIC "-//freedesktop//DTD D-BUS Object Introspection 1.0//EN"
+"http://www.freedesktop.org/standards/dbus/1.0/introspect.dtd">
+<node>
+
+<interface name="org.freedesktop.DBus.Introspectable">
+
+<method name="Introspect"><arg name="xml" type="s" direction="out"/>
+</method>
+
+</interface>
+
+<!-- The name of the sailfish connman iptables plugin interface -->
+<interface name="org.sailfishos.connman.mdm.iptables">
+
+<!-- Each function returns a code and descriptive string.
+The returned result codes and string pairs are:
+0	"Ok",
+1	"Invalid IP",
+2	"Invalid port",
+3	"Invalid port range",
+4	"Invalid service name",
+5	"Invalid protocol",
+6	"Invalid policy",
+7	"Invalid file path",
+8	"Cannot process rule",
+9	"Cannot perform operation",
+-->
+
+<!-- 
+Clears given table from iptables. Defaults to filter table.
+Sends signal FirewallCleared
+-->
+<method name="ClearIptablesTable">
+<arg name="table" type="s" direction="in"/>
+<arg name="result" type="q" direction="out"/>
+<arg name="string" type="s" direction="out"/>
+</method>
+
+<!--
+Change policy of filter table INPUT chain. Parameter is the same as iptables
+command (ACCEPT/DROP)
+Sends signal PolicyChanged
+-->
+<method name="ChangeInputPolicy">
+<arg name="policy" type="s" direction="in"/>
+<arg name="result" type="q" direction="out"/>
+<arg name="string" type="s" direction="out"/>
+</method>
+
+<!--
+Change policy of filter table OUTPUT chain. Parameter is the same as iptables
+command (ACCEPT/DROP)
+Sends signal PolicyChanged
+-->
+<method name="ChangeOutputPolicy">
+<arg name="policy" type="s" direction="in"/>
+<arg name="result" type="q" direction="out"/>
+<arg name="string" type="s" direction="out"/>
+</method>
+
+<!--
+Following methods take following parameters as input:
+ NAME		TYPE	DESCRIPTION
+ ip		s		IP address or network with CIDR or dot notation (1.2.3.4/24
+ 					or 1.2.3.4/255.255.255.0)
```

Comment:
IPv4 or IPv6? Number of mask bits should be a separate parameter, I think. Also, it makes every bit of sense to send IP address as a byte array to avoid ambiguities associated with text representation (like `012.0.0.1` vs `10.0.0.1` - is it the same address or not? not to mention various kinds of IPv6 representation). If both IPv4 and IPv6 are supposed to be supported, I would add address family as a separate parameter (although some could argue that one can tell IPv4 from IPv6 based on number of bytes in the address).

Reply:
Only IPv4 is currently required, some parts were thought IPv6 in mind, excluding the API. 

Here is a following problem with byte-encoded IP-addresses: The app using the API has to use some lists of byte-encoded addresses to be sent to the API, which are not human readable, thus, such data is not feasible to be added into a configuration file. Binary storage of the addresses works of course, but the content is not directly editable. Each edit would require a conversion from byte-encoded format to text format. Same thing happens with the API use: the application using the API would send the IP addresses as byte-encoded and the validity can be checked if the data is compatible with `sockaddr_in/sockaddr_in6` structures.

The sockaddr_in contains address as `uint32`, whereas in `sockaddr_in6` address is `unsigned char[16]`. It would, therefore need separate methods. Unless we want the `sockaddr_storage`structure to be sent, which may result in taking care of  the struct.

I think here a standard way is the best one, so either support the address as string or separate methods for the two IP types. 

Eventually the API has to convert the address to string to be used in iptables. IPv4 and IPv6 addresses are easy to differentiate, hence the different notation. Yes, with IPv6 some parts of the address can be omitted and should be treated as zeros, and there are IPv4-mapped IPv6 addresses to take care of. We can surely skip link local IPv6 and localhost addresses.

So it is a question about which part should do the conversion. The way I see that this plugin as an API is the place where input has to be verified and sanitized (also for leading white space) and not to rely on anything the application sends. Surely with string formatted addresses there is more work on the plugin. Other parameters can be changed to integers in next API version.

And a good point about the address with leading zeroes, these seem to pass through `getaddrinfo()` without errors. Need to check this.

Reply:
And also in https://tools.ietf.org/html/draft-main-ipaddr-text-rep-02 it is defined that: 

> 3.1 IPv4 Dotted Octet Format
>   A 32-bit IPv4 address is divided into four octets.  Each octet is represented numerically in decimal, using the minimum possible number of digits (leading zeroes are not used, except in the case of 0 itself).

So technically `012.0.0.1` is invalid input and should be treated as such.

Reply:
Added check of leading zeros in IPv4 addresses: f4aafb6587fb3396516ad38ffdbc4fa50d654dc5